PKCE Support

mhum opened the pull request:
This is just a draft PR, but I think it's feature complete. I just want to add some more tests for the other parts of the code I touched.

The RFC gives the option of storing the code within the server or as part of the `code` that is returned from the authorization request. That seems weird to me, tbh. Since the authorization codes are short-lived in stored in our DB, I figured it was easiest just to store the code challenge, too.

evert commented on src/oauth2/service.ts:
```diff
@@ -142,7 +142,7 @@ export async function generateTokenForClient(client: OAuth2Client): Promise<OAut
  * The resource owner then exchanges that code for an access and refresh token.
  */
-export async function generateTokenFromCode(client: OAuth2Client, code: string): Promise<OAuth2Token> {
+export async function generateTokenFromCode(client: OAuth2Client, code: string, codeVerifier: string|undefined): Promise<OAuth2Token> {
```
I think it might be nice to make `codeVerifier` get a more specific type like `'S256' | 'plain'`, and make sure that this is validated as well. (earlier, in the controller)

mhum replied:
Here, `codeVerifier` refers to the verification code the client sends in the body of the token request _after_ the authorization request happens (where the challenge method is provided). This is only ever a string (if provided) or undefined (if not provided).

But this does bring up the fact that I'm not validating the `codeChallengeMethod` in the authorization request. I did think of that later. I forgot to add _any_ validation to the authorization request (such as providing the `codeChallenge` but not the `codeChallengeMethod`). I'll add a check there to verify that the provided `codeChallengeMethod` (if included) is either `plain` or `S256`.

evert replied:
Cool!

evert commented on src/oauth2/service.ts:
```diff
@@ -169,6 +171,33 @@ export async function generateTokenFromCode(client: OAuth2Client, code: string):
+  }
+
+  // For the plain method, the derived code and the code verifier are the same
+  let derivedCodeChallenge = codeVerifier;
```
Is `plain` secure? I don't really know the PKCE spec too well!

mhum replied:
No, it's to accommodate clients that cannot do sha256 for technical reasons since PKCE is required in 2.1.

evert replied:
I see. I remember one of the major issues with JWT was that many libraries by default also supported a 'none' encryption method. So I hope this is not a similar thing, but I would guess not.

mhum replied:
I would hope oauth libraries that support sha256 wouldn't provide the implementor the option of disabling it. The RFC does say the client MUST use sha256 if it can technically.
https://tools.ietf.org/html/draft-parecki-oauth-v2-1-01#section-4.1.1.2

evert commented on src/oauth2/service.ts:
```diff
@@ -169,6 +171,33 @@ export async function generateTokenFromCode(client: OAuth2Client, code: string):
 
 }
 
+export function validatePKCE(codeVerifier: string|undefined, codeChallenge: string|undefined, codeChallengeMethod: string|undefined) {
```
Could we simply *not* call this function when PKCE is not set up? I feel a bit better about that than to make every argument potentially undefined. This means that this check has to move to the call-site

mhum replied:
I guess I can. Feels a bit weird to me to have something like this:
```
if (codeVerifier !! codeChallenge !! codeChallengeMethod) {
  validatePKCE(codeVerifier, codeChallenge, codeChallengeMethod)
}

Since PKCE is required by 2.1 and supported in 2.0, every token request can potentially require PKCE validation. It makes more logical sense to me for that to be the first thing to do while doing PKCE validation (do we even have to do PKCE validation). I don't like separating the "do we need to do PKCE validation" and the "do PKCE validation" since it's potentially applicable to every request.

evert replied:
I think the risk of also allowing undefined for any of the parameters is that it gets easier to slip in bugs, and call that function with a bad argument.

If the function requires a string, the caller is forced to ensure that it really is a string and if it isn't, get a hard error. I feel that this makes the source a bit more resilient against potential bugs.

You might also consider passing in something else, like an object with all the properties. An issue here too is that given each of the arguments may or may not be undefined, this creates 8 different branches, whereas a type like this:

```
type PkceChallenge = {
  codeVerifier: string,
  codeChallenge: string,
  codeChallengeMethod: 'plain' | 'S256'
} | undefined
```
Only creates 2 branches, and it's a bit more obvious what the behavior of the 2 branches will be.

evert replied:
Third alternative is to use overloads:

```typescript
 export function validatePKCE(codeVerifier: string, codeChallenge: string, codeChallengeMethod: string) 
 export function validatePKCE(codeVerifier: undefined, codeChallenge: undefined, codeChallengeMethod: undefined);
 export function validatePKCE(codeVerifier: string|undefined, codeChallenge: string|undefined, codeChallengeMethod: string|undefined) { 
```

This enforces that all arguments are either all undefined, or all strings.

mhum replied:
If I don't make the parameters optional or provide an object where the fields are optional, I can't do validation if the user does something wrong. 

For example, the user provides a challenge code in the authorization request but doesn't provide the verifier code in the token request (in which case `codeVerifier` would be missing). Or if the user provides the verified code in the token request, but didn't provide the challenge code + method in the authorization request (so `codeChallenge` and `codeChallengeMethod` would be missing). 

I could push all of that parameter validation up into `generateTokenFromCode`, but that's a lot of logic specific to doing PKCE validation inside a function meant to generate tokens.

evert replied:
Fair, that sounds good!

mhum commented:
This is ready for review. Changes since the initial draft:
- Param validation on authorization get/post requests
- Add a `start:watch` script for easy developing (by adding `tsc-watch`)
- Fixed a couple of `log` calls that were missing parameters
- Small spelling fix
- Added `chai-as-promised` for tests

evert commented on src/oauth2/controller/authorize.ts:
```diff
@@ -190,11 +212,20 @@ class AuthorizeController extends Controller {
+    redirectUri: string,
+    state: string|undefined,
+    codeChallenge: string|undefined,
+    codeChallengeMethod: string|undefined,
```
Could this type be refined to `'S256' | 'plain' | undefined` vs just `string` ?

mhum replied:
Is that valid syntax? I can't find anything in the TS docs about doing that. When I try it, I get an error on the calling side:
```
Argument of type 'string' is not assignable to parameter of type '"plain" | "S256"'.
```

mhum replied:
Got it. Combo of string literals
https://www.typescriptlang.org/docs/handbook/advanced-types.html#string-literal-types
and type assertions
https://www.typescriptlang.org/docs/handbook/basic-types.html#type-assertions

evert replied:
Basically, we want the strictest possible types for everything.

To fix the error, you also need to type the original definition the same way:

   let codeChallengeMethod: 'plain' | 'S256' | undefined;

It might also help to create a new reuable type for the enum. Ideally we use this everywhere this is applicable.

    type CodeChallengeMethod = 'plain' | 'S256';

The benefit if taking this approach and validating values *before* they enter the function, is that the function can only ever be called with a correct value, which means no validation in the function is needed.

I wrote a bit about this here:

https://evertpot.com/typescript-changes-code/

The official docs are here:

https://www.typescriptlang.org/docs/handbook/advanced-types.html#string-literal-types

mhum replied:
The validation still needs to happen _somewhere_ since these are user provided values. Even though TS will now prevent us from calling this function with a different value, validation now just needs to happen before calling the function.

evert commented on src/oauth2/controller/authorize.ts:
```diff
@@ -43,7 +43,7 @@ class AuthorizeController extends Controller {
     const redirectUri = ctx.query.redirect_uri;
     const codeChallenge = ctx.query.code_challenge;
-    const codeChallengeMethod = ctx.query.code_challenge_method ? ctx.query.code_challenge_method : 'plain';
+    const codeChallengeMethod = ctx.query.code_challenge_method ? (<'S256'>ctx.query.code_challenge_method) : 'plain';
```
This is effectively a type cast, if the code_challenge_method is garbage, it will be typed as `S256`.

Ideal I think would be a strict check. Error when it's not one of the 2 acceptable values.

Bit more verbose, but:

```suggestion
let codeChallengeMethod: CodeChallengeMethod = undefined;
switch(ctx.query.code_challenge_method) {
  case: 'S256':
  case 'plain' :
    codeChallengeMethod = ctx.query.code_challenge_method;
    break;
  default:
    throw new Error('...');
}
```

mhum replied:
I'm effectively doing the same check on line 33:
https://github.com/curveball/a12n-server/blob/29e05d92d4594f352cff53cc23ffab4150bff79a/src/oauth2/controller/authorize.ts#L33

evert replied:
Maybe you can use an assertion function for easy re-use?

```
function assertIsChallengeMethod(input: string): asserts input is CodeChallengeMethod {

  if (input !== 'plain' && input !=='S256') {
    throw new Error('...');
  }
}
```

Used as such:

```
let codeChallengeMethod: CodeChallengeMethod | undefined;
if (ctx.query.code_challenge_method) {
   assertsIsChallengeMethod(ctx.query.code_challenge_method);
   codeChallengeMethod = ctx.query.code_challenge_method;
}
```

evert replied:
An assertion function is 'trusted' by typescript to do the correct thing. Typescript will assume that you _will_ throw an error if the input was not correct.

After an assertion function is called, typescript will internally have narrowed down the `string` to `CodeChallengeMethod`.

mhum replied:
I created a type and went with the switch statement.

evert commented on src/oauth2/service.ts:
```diff
@@ -169,6 +171,33 @@ export async function generateTokenFromCode(client: OAuth2Client, code: string):
+  // For the plain method, the derived code and the code verifier are the same
+  let derivedCodeChallenge = codeVerifier;
+
+  if (codeChallengeMethod === 'S256') {
```
This is a suggestion, not a must... but if this were to be done with a switch (that also handles `plain`) you could use this approach:

https://stackoverflow.com/a/58009992/80911

to ensure that Typescript always complains if the enum ever gets expanded with a third "method".

The switch with `never` type is a bit ugly, so another option is to move this to a function that takes a `codeChallengeMethod`, a `codeVerifier` and returns the `derivedCodeChallenge`.

mhum replied:
I feel like a switch statement is overkill for only two possibilities. If there was a third type of challenge method, I'd create a separate function. I guess I could have done this with a ternary, too.

evert replied:
The suggestion has more to related to getting a static error check once a third option is added. Basically when `CodeChallengeMethod` is altered with a new option, in the ideal case typescript refuses to compile it until that new case is explicitly handled. Switch is one way to handle this.. but this is not a big deal.

BTW I know other code in this repo is _also_ not really guarded as well and written this way.. I changed how I write things quite a bit since I wrote this stuff =)

Anyway, this works!

mhum commented:
Resolves #168
